File: backend/services/player_service.py
"""Player Service - Business logic for player operations"""

from typing import List, Optional, Dict, Tuple
import logging
from datetime import datetime

logger = logging.getLogger(__name__)


class PlayerService:
    """Service class for player-related operations"""
    
    def __init__(self):
        """Initialize player service"""
        self.cache = {}
        logger.info("✅ PlayerService initialized")
    
    @staticmethod
    def get_all_players(
        page: int = 1,
        per_page: int = 20,
        filters: Optional[Dict] = None
    ) -> Dict:
        """Get all players with pagination and filters"""
        
        logger.info(f"Getting players - Page: {page}, Per Page: {per_page}")
        
        # TODO: Implement database query with filters
        # Filters could include: team, position, league, age, etc.
        
        return {
            'total': 18000,
            'page': page,
            'per_page': per_page,
            'filters': filters or {},
            'data': []
        }
    
    @staticmethod
    def get_player_by_id(player_id: int) -> Optional[Dict]:
        """Get specific player details by ID"""
        
        logger.info(f"Fetching player: {player_id}")
        
        # TODO: Query database for player details
        
        player = {
            'id': player_id,
            'name': 'Player Name',
            'age': 25,
            'position': 'Midfielder',
            'team': 'Team Name',
            'nationality': 'Country',
            'jersey_number': 10,
            'height': '180 cm',
            'weight': '75 kg',
            'foot': 'Right',
            'contract_until': '2026'
        }
        
        return player
    
    @staticmethod
    def search_players(query: str, limit: int = 20) -> List[Dict]:
        """Search players by name or criteria"""
        
        logger.info(f"Searching players with query: {query}")
        
        # TODO: Implement fuzzy search or full-text search
        
        return []
    
    @staticmethod
    def get_player_stats(player_id: int) -> Dict:
        """Get comprehensive player statistics"""
        
        logger.info(f"Getting stats for player: {player_id}")
        
        # TODO: Calculate stats from match data
        
        stats = {
            'player_id': player_id,
            'appearances': 0,
            'goals': 0,
            'assists': 0,
            'yellow_cards': 0,
            'red_cards': 0,
            'passes_completed': 0,
            'pass_success_rate': 0.0,
            'tackles': 0,
            'interceptions': 0,
            'fouls_committed': 0,
            'fouls_drawn': 0,
            'average_rating': 0.0,
            'minutes_played': 0,
            'goals_per_game': 0.0,
            'assists_per_game': 0.0
        }
        
        return stats
    
    @staticmethod
    def compare_players(player_ids: List[int]) -> Dict:
        """Compare multiple players"""
        
        logger.info(f"Comparing players: {player_ids}")
        
        # TODO: Get stats for each player and create comparison
        
        comparison = {
            'players': player_ids,
            'stats': [],
            'comparison_table': {}
        }
        
        return comparison
    
    @staticmethod
    def get_top_scorers(limit: int = 10, league: Optional[str] = None) -> List[Dict]:
        """Get top scorers in league"""
        
        logger.info(f"Getting top {limit} scorers" + (f" in {league}" if league else ""))
        
        # TODO: Query and sort by goals
        
        return []
    
    @staticmethod
    def get_top_assists(limit: int = 10, league: Optional[str] = None) -> List[Dict]:
        """Get top assist providers"""
        
        logger.info(f"Getting top {limit} assist providers")
        
        # TODO: Query and sort by assists
        
        return []
    
    @staticmethod
    def get_player_career_stats(player_id: int) -> Dict:
        """Get player career statistics"""
        
        logger.info(f"Getting career stats for player: {player_id}")
        
        # TODO: Aggregate all season stats
        
        career_stats = {
            'player_id': player_id,
            'total_appearances': 0,
            'total_goals': 0,
            'total_assists': 0,
            'seasons': [],
            'best_season': {}
        }
        
        return career_stats
    
    @staticmethod
    def get_players_by_position(position: str, limit: int = 20) -> List[Dict]:
        """Get players by position"""
        
        logger.info(f"Getting {limit} {position} players")
        
        # TODO: Query by position filter
        
        return []
    
    @staticmethod
    def get_players_by_team(team_id: int) -> List[Dict]:
        """Get all players in a team"""
        
        logger.info(f"Getting players for team: {team_id}")
        
        # TODO: Query players by team_id
        
        return []
    
    @staticmethod
    def get_player_form(player_id: int, last_matches: int = 5) -> Dict:
        """Get player recent form"""
        
        logger.info(f"Getting form for player: {player_id}")
        
        # TODO: Analyze recent performances
        
        form = {
            'player_id': player_id,
            'recent_matches': [],
            'average_rating': 0.0,
            'trend': 'stable'  # improving, declining, stable
        }
        
        return form
    
    @staticmethod
    def get_player_consistency(player_id: int) -> Dict:
        """Get player consistency metrics"""
        
        logger.info(f"Calculating consistency for player: {player_id}")
        
        # TODO: Calculate standard deviation of ratings
        
        consistency = {
            'player_id': player_id,
            'consistency_score': 0.0,
            'performance_variance': 0.0
        }
        
        return consistency

File: frontend/src/services/playerService.ts
import { apiClient } from './api'

/**
 * Player Service
 * Handles all player-related API calls
 */

export interface Player {
  id: number
  name: string
  age: number
  nationality: string
  position: string
  team_id: number
  team_name?: string
  jersey_number: number
  height: string
  weight: string
  foot: string
  goals: number
  assists: number
  appearances: number
  rating: number
  pass_accuracy: number
  tackles: number
  yellow_cards: number
  red_cards: number
  fouls_committed: number
  created_at: string
  updated_at: string
}

export interface PlayerStats {
  player_id: number
  goals: number
  assists: number
  appearances: number
  rating: number
  pass_accuracy: number
  tackles: number
  interceptions: number
  yellow_cards: number
  red_cards: number
}

export interface PlayerResponse {
  status: string
  data: Player | Player[]
  message: string
}

class PlayerService {
  /**
   * Get all players
   */
  static async getPlayers(page = 1, perPage = 20, filters?: any) {
    try {
      const response = await apiClient.get<PlayerResponse>('/players', {
        params: {
          page,
          per_page: perPage,
          ...filters,
        },
      })
      return response.data
    } catch (error) {
      console.error('Error fetching players:', error)
      throw error
    }
  }

  /**
   * Get player by ID
   */
  static async getPlayerById(playerId: number) {
    try {
      const response = await apiClient.get<PlayerResponse>(`/players/${playerId}`)
      return response.data
    } catch (error) {
      console.error(`Error fetching player ${playerId}:`, error)
      throw error
    }
  }

  /**
   * Search players
   */
  static async searchPlayers(query: string, limit = 20) {
    try {
      const response = await apiClient.get<PlayerResponse>('/players/search', {
        params: { q: query, limit },
      })
      return response.data
    } catch (error) {
      console.error('Error searching players:', error)
      throw error
    }
  }

  /**
   * Get player statistics
   */
  static async getPlayerStats(playerId: number) {
    try {
      const response = await apiClient.get<PlayerResponse>(`/players/${playerId}/stats`)
      return response.data
    } catch (error) {
      console.error(`Error fetching player stats for ${playerId}:`, error)
      throw error
    }
  }

  /**
   * Get top scorers
   */
  static async getTopScorers(limit = 10) {
    try {
      const response = await apiClient.get<PlayerResponse>('/players/top-scorers', {
        params: { limit },
      })
      return response.data
    } catch (error) {
      console.error('Error fetching top scorers:', error)
      throw error
    }
  }

  /**
   * Get top assists
   */
  static async getTopAssists(limit = 10) {
    try {
      const response = await apiClient.get<PlayerResponse>('/stats/top-assists', {
        params: { limit },
      })
      return response.data
    } catch (error) {
      console.error('Error fetching top assists:', error)
      throw error
    }
  }

  /**
   * Get players by team
   */
  static async getPlayersByTeam(teamId: number) {
    try {
      const response = await apiClient.get<PlayerResponse>(`/teams/${teamId}/players`)
      return response.data
    } catch (error) {
      console.error(`Error fetching players for team ${teamId}:`, error)
      throw error
    }
  }

  /**
   * Compare players
   */
  static async comparePlayers(playerIds: number[]) {
    try {
      const response = await apiClient.post<PlayerResponse>('/players/compare', {
        player_ids: playerIds,
      })
      return response.data
    } catch (error) {
      console.error('Error comparing players:', error)
      throw error
    }
  }

  /**
   * Get player form
   */
  static async getPlayerForm(playerId: number, lastMatches = 5) {
    try {
      const response = await apiClient.get<PlayerResponse>(`/players/${playerId}/form`, {
        params: { last_matches: lastMatches },
      })
      return response.data
    } catch (error) {
      console.error(`Error fetching player form for ${playerId}:`, error)
      throw error
    }
  }

  /**
   * Get player efficiency rating
   */
  static async getPlayerEfficiency(playerId: number) {
    try {
      const response = await apiClient.get<PlayerResponse>(`/players/${playerId}/efficiency`)
      return response.data
    } catch (error) {
      console.error(`Error fetching player efficiency for ${playerId}:`, error)
      throw error
    }
  }

  /**
   * Get trending players
   */
  static async getTrendingPlayers(limit = 10) {
    try {
      const response = await apiClient.get<PlayerResponse>('/stats/trending', {
        params: { limit },
      })
      return response.data
    } catch (error) {
      console.error('Error fetching trending players:', error)
      throw error
    }
  }
}

export default PlayerService
